### PR TITLE
Use the new rustc interface

### PR DESCRIPTION
**Pull request by oli-obk**
rebase of https://github.com/rust-lang/miri/pull/608

### CODE REVIEW

File: src/bin/miri.rs [hunk 1]
```diff
@@ -11,115 +11,46 @@ extern crate rustc_metadata;
...
-        codegen_backend: &CodegenBackend,
-        matches: &getopts::Matches,
-        sess: &Session,
-        cstore: &CStore,
-        input: &Input,
-        odir: &Option<PathBuf>,
-        ofile: &Option<PathBuf>,
-    ) -> Compilation {
-        // Called *before* `build_controller`. Add filename to `miri` arguments.
-        self.miri_config.args.insert(0, input.filestem().to_string());
```
Comment (RalfJung) [inline src/bin/miri.rs]: Seems like this line got entirely lost?

**Comment by RalfJung** [comment]
Rerunning doesn't help, you need to bump the Rust version.

**Comment by oli-obk** [comment]
Hmm... I'm getting a local failure:

```
error: error pattern ' reached the configured maximum number of stack frames' not found!
status: exit code: 1
command: "target/debug/miri" "tests/compile-fail/stack_limit.rs" "-L" "/tmp/compiletestDVKpG1" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/tmp/compiletestDVKpG1/stack_limit.stage-id" "-Dwarnings" "-Dunused" "--edition" "2018" "-L" "/tmp/compiletestDVKpG1/stack_limit.stage-id.aux" "-A" "unused"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error: unused attribute
 --> tests/compile-fail/stack_limit.rs:2:1
  |
2 | #![miri(stack_limit=16)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-attributes` implied by `-D unused`

error: aborting due to previous error
```

I probably lost marking these attributes as used in the rebase...

**Comment by RalfJung** [comment]
On CI not even the build completes:
```
error[E0107]: wrong number of lifetime arguments: expected 0, found 1
  --> src/bin/miri-rustc-tests.rs:35:64
   |
35 |     fn after_parsing(&mut self, compiler: &interface::Compiler<'_>) -> bool {
   |                                                                ^^ unexpected lifetime argument
error[E0107]: wrong number of lifetime arguments: expected 0, found 1
  --> src/bin/miri-rustc-tests.rs:46:65
   |
46 |     fn after_analysis(&mut self, compiler: &interface::Compiler<'_>) -> bool {
   |                                                                 ^^ unexpected lifetime argument
```

**Comment by oli-obk** [comment]
I can't find any code related to the resource limits... seems like we haven't had those for a while.

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/miri/issues/643

**Comment by RalfJung** [comment]
There are still some warnings:
```
warning: unused `std::result::Result` that must be used
  --> benches\helpers\miri_helper.rs:62:5
   |
62 |     rustc_driver::run_compiler(args, &mut MiriCompilerCalls { bencher }, None, None);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_must_use)] on by default
   = note: this `Result` may be an `Err` variant, which should be handled
warning: unused `std::result::Result` that must be used
  --> benches\helpers\miri_helper.rs:62:5
   |
62 |     rustc_driver::run_compiler(args, &mut MiriCompilerCalls { bencher }, None, None);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_must_use)] on by default
   = note: this `Result` may be an `Err` variant, which should be handled
warning: unused `std::result::Result` that must be used
  --> benches\helpers\miri_helper.rs:62:5
   |
62 |     rustc_driver::run_compiler(args, &mut MiriCompilerCalls { bencher }, None, None);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_must_use)] on by default
   = note: this `Result` may be an `Err` variant, which should be handled
```

**Pull request merged**